### PR TITLE
fix potential memory leak in jni code

### PR DESCRIPTION
**Pull request by zaki50**
Added missing `ReleaseLongArrayElements` and `ReleaseBooreanArrayElements` call.

please review this @realm/java

### CODE REVIEW

**Comment by zaki50** [comment]
retest this please

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 1]
```diff
@@ -70,6 +70,9 @@ JNIEXPORT jstring JNICALL Java_io_realm_internal_TableQuery_nativeValidateQuery
 #define RELEASE_ARRAY() \
     env->ReleaseLongArrayElements(columnIndexes, arr, 0);
 
+#define ABORT_ARRAY() \
+    env->ReleaseLongArrayElements(columnIndexes, arr, JNI_ABORT);
+
```
Comment (beeender) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: 0 copy back the content and free the elems buffer
JNI_COMMIT copy back the content but do not free the elems buffer
JNI_ABORT free the buffer without copying back the possible changes

I don't think even for `RELEASE_ARRAY` we need to copy anything back. Shall we just change the `RELEASE_ARRAY` to use `JNI_ABORT`?

**Comment by beeender** [comment]
Maybe it is the time we refactor this into something like below? Then no need to care about the release anymore, even when exception thrown.

```
class JniArray<T, T_ARRAY> {
public:
    JniArray(env, T_ARRAY columnIndexes) :
        env(env),
        arr_len(env->GetArrayLength(columnIndexes)),
        arr(env->GetLongArrayElements(columnIndexes, NULL)){
        
    }
    ~JniArray() {
        env->ReleaseLongArrayElements(columnIndexes, arr, JNI_ABORT);
    }
    const T* arr;
    const jsize arr_len;
    const Env* env;
}
```

**Comment by zaki50** [comment]
@beeender good idea.

We need a feature to choose  `0`(copy from native array to java array) and `JNI_ABORT `(no need to write back to java array) when releasing.
In most (all?) cases, we don't modify the array though.

**Comment by zaki50** [comment]
Oh, I didn't read the first comment.
We might not need to write back to java array.

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp (hunk 1 above)
Comment (zaki50) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: I agree with you.

I'll change that.

Comment (zaki50) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: all tests passed. I'll push that.

**Comment by kneth** [comment]
@beeender Can be do generic enough? That is, for all `T` and `T_ARRAY` and not only `jlong`?

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 2]
```diff
@@ -128,7 +128,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeEqual__J_3JJ(
     try {
         if (arr_len == 1) {
             if (!QUERY_COL_TYPE_VALID(env, nativeQueryPtr, arr[0], type_Int))
+            {
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: I think a opening braces should not be splited as seperated lines.

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 3]
```diff
@@ -146,7 +149,10 @@ JNIEXPORT void JNICALL JNICALL Java_io_realm_internal_TableQuery_nativeNotEqual_
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 4]
```diff
@@ -164,7 +170,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeGreater__J_3JJ(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 5]
```diff
@@ -182,7 +191,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeGreaterEqual__J_3
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 6]
```diff
@@ -200,7 +212,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeLess__J_3JJ(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 7]
```diff
@@ -218,7 +233,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeLessEqual__J_3JJ(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 8]
```diff
@@ -235,7 +253,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeBetween__J_3JJJ(
     GET_ARRAY()
     if (arr_len == 1) {
         if (!QUERY_COL_TYPE_VALID(env, nativeQueryPtr, arr[0], type_Int))
+        {
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 9]
```diff
@@ -255,7 +276,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeEqual__J_3JF(
     try {
         if (arr_len == 1) {
             if (!QUERY_COL_TYPE_VALID(env, nativeQueryPtr, arr[0], type_Float))
+            {
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 10]
```diff
@@ -273,7 +297,10 @@ JNIEXPORT void JNICALL JNICALL Java_io_realm_internal_TableQuery_nativeNotEqual_
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 11]
```diff
@@ -291,7 +318,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeGreater__J_3JF(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 12]
```diff
@@ -309,7 +339,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeGreaterEqual__J_3
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 13]
```diff
@@ -327,7 +360,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeLess__J_3JF(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 14]
```diff
@@ -345,7 +381,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeLessEqual__J_3JF(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 15]
```diff
@@ -363,7 +402,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeBetween__J_3JFF(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 16]
```diff
@@ -383,7 +425,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeEqual__J_3JD(
     try {
         if (arr_len == 1) {
             if (!QUERY_COL_TYPE_VALID(env, nativeQueryPtr, arr[0], type_Double))
+            {
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 17]
```diff
@@ -401,7 +446,10 @@ JNIEXPORT void JNICALL JNICALL Java_io_realm_internal_TableQuery_nativeNotEqual_
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 18]
```diff
@@ -419,7 +467,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeGreater__J_3JD(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 19]
```diff
@@ -437,7 +488,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeGreaterEqual__J_3
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 20]
```diff
@@ -455,7 +509,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeLess__J_3JD(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 21]
```diff
@@ -473,7 +530,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeLessEqual__J_3JD(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 22]
```diff
@@ -491,7 +551,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeBetween__J_3JDD(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 23]
```diff
@@ -511,7 +574,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeEqualDateTime(
     try {
         if (arr_len == 1) {
             if (!QUERY_COL_TYPE_VALID(env, nativeQueryPtr, arr[0], type_DateTime))
+            {
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 24]
```diff
@@ -530,7 +596,10 @@ JNIEXPORT void JNICALL JNICALL Java_io_realm_internal_TableQuery_nativeNotEqualD
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 25]
```diff
@@ -549,7 +618,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeGreaterDateTime(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 26]
```diff
@@ -567,7 +639,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeGreaterEqualDateT
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 27]
```diff
@@ -585,7 +660,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeLessDateTime(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 28]
```diff
@@ -603,7 +681,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeLessEqualDateTime
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 29]
```diff
@@ -621,7 +702,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeBetweenDateTime(
...
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 30]
```diff
@@ -640,7 +724,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeEqual__J_3JZ(
     try {
         if (arr_len == 1) {
             if (!QUERY_COL_TYPE_VALID(env, nativeQueryPtr, arr[0], type_Bool))
+            {
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 31]
```diff
@@ -673,7 +760,10 @@ void TableQuery_StringPredicate(JNIEnv *env, jlong nativeQueryPtr, jlongArray co
         JStringAccessor value2(env, value); // throws
         if (arr_len == 1) {
             if (!QUERY_COL_TYPE_VALID(env, nativeQueryPtr, arr[0], type_String))
+            {
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 32]
```diff
@@ -1430,7 +1536,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeIsNull(
         jlong column_idx = arr[arr_len-1];
         TableRef table_ref = getTableByArray(nativeQueryPtr, arr, arr_len);
         if (!TBL_AND_COL_NULLABLE(env, table_ref.get(), column_idx))
+        {
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_TableQuery.cpp [hunk 33]
```diff
@@ -1572,7 +1683,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeIsNotNull
         TableRef table_ref = getTableByArray(nativeQueryPtr, arr, arr_len);
 
         if (!TBL_AND_COL_NULLABLE(env, table_ref.get(), column_idx))
+        {
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-jni/src/io_realm_internal_table.cpp [hunk 34]
```diff
@@ -1390,7 +1390,11 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_Table_nativeGetSortedViewMulti(
 
     for (int i = 0; i < arr_len; ++i) {
         if (!TBL_AND_COL_INDEX_VALID(env, pTable, S(long_arr[i]) ))
+        {
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_table.cpp]: same

File: realm/realm-jni/src/io_realm_internal_tableview.cpp [hunk 35]
```diff
@@ -911,12 +908,19 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableView_nativeSortMulti(
...
 
+        jlong *long_arr = env->GetLongArrayElements(columnIndices, NULL);
+        jboolean *bool_arr = env->GetBooleanArrayElements(ascending, NULL);
+
         std::vector<size_t> indices;
         std::vector<bool> ascendings;
 
         for (int i = 0; i < arr_len; ++i) {
             if (!COL_INDEX_VALID(env, TV(nativeViewPtr), long_arr[i]))
+            {
```
Comment (dalinaum) [inline realm/realm-jni/src/io_realm_internal_tableview.cpp]: same

**Comment by zaki50** [comment]
@beeender I gave up changing `Get*ArrayElements` and `Release*ArrayElements` according to array type.

How about the following code?

```
class JniLongArray {
public:
    JniLongArray(JNIEnv* env, jlongArray javaArray);

    ~JniLongArray();

    jsize length() const noexcept;

    jlong* ptr() const noexcept;

    jlong& operator[](const int index) noexcept;

    void updateOnRelease() noexcept;

private:
    JNIEnv*    const m_env;
    jlongArray const m_javaArray;
    jsize      const m_arrayLength;
    jlong*     const m_array;
    jint             m_releaseMode;
};
```

```
JniLongArray::JniLongArray(JNIEnv* env, jlongArray javaArray)
        : m_env(env)
        , m_javaArray(javaArray)
        , m_arrayLength(env->GetArrayLength(javaArray))
        , m_array(env->GetLongArrayElements(javaArray, NULL))
        , m_releaseMode(JNI_ABORT) {
}

JniLongArray::~JniLongArray() {
    m_env->ReleaseLongArrayElements(m_javaArray, m_array, m_releaseMode);
}

jsize JniLongArray::len() const noexcept
{
    return m_arrayLength;
}

jlong* JniLongArray::ptr() const noexcept
{
    return m_array;
}

jlong& JniLongArray::operator[] (const int index) noexcept
{
    return m_array[index];
}

void JniLongArray::updateOnRelease() noexcept
{
    m_releaseMode = 0;
}
```

```
JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeEqual__J_3JJ(
    JNIEnv* env, jobject, jlong nativeQueryPtr, jlongArray columnIndexes, jlong value)
{
    JniLongArray arr(env, columnIndexes);
    try {
        if (arr.len() == 1) {
            if (!QUERY_COL_TYPE_VALID(env, nativeQueryPtr, arr[0], type_Int)) {
                return;
            }
            Q(nativeQueryPtr)->equal(S(arr[0]), static_cast<int64_t>(value));
        }
        else {
            TableRef table_ref = getTableForLinkQuery(nativeQueryPtr, arr.ptr(), arr.len());
            Q(nativeQueryPtr)->and_query(numeric_link_equal<Int, int64_t, jlong>(table_ref, arr[arr.len()-1], value));
        }
    } CATCH_STD()
}
```

**Comment by zaki50** [comment]
updated.
review please @realm/java

**Comment by dalinaum** [comment]
Besides a minor comment, LGTM :+1:

**Comment by zaki50** [comment]
I refactored helper methods.

review this please @realm/java

**Comment by beeender** [comment]
@kneth Failed use template to make it generic :( since we need to call different methods to free the array `jniReleaseLongArray` vs `jniReleaseBooleanArray`

File: realm/realm-jni/src/util.cpp [hunk 36]
```diff
@@ -411,3 +411,75 @@ JStringAccessor::JStringAccessor(JNIEnv* env, jstring str)
...
+        , m_arrayLength(env->GetArrayLength(javaArray))
+        , m_array(env->GetLongArrayElements(javaArray, NULL))
+        , m_releaseMode(JNI_ABORT) {
+}
+
+JniLongArray::~JniLongArray() {
+    m_env->ReleaseLongArrayElements(m_javaArray, m_array, m_releaseMode);
+}
+
+jsize JniLongArray::len() const noexcept
```
Comment (kneth) [inline realm/realm-jni/src/util.cpp]: Consider to `inline` these short methods.

**Comment by kneth** [comment]
:+1: with a minor performance suggestion.

**Comment by zaki50** [comment]
inlined all short methods

**Comment by beeender** [comment]
:+1:

**Comment by zaki50** [comment]
thank you for reviewing. I'll squash and merge this PR.

**Pull request merged**